llvm/lib/IR/RepoTicket.cpp

```diff
@@ -163,91 +143,89 @@ updateDigestUseCallDependencies(const GlobalObject *GO, MD5 &GOHash,
 
   bool Changed = false;
   GOHash.update('T');
-  auto State = UpdateDigestAndGetDependencies(GO, GOHash, GOIMap);
-  Changed = Changed || State.second;
-
-  // Recursively for all the dependent global objects.
-  for (const GlobalObject *D : State.first) {
-    const llvm::Function *Fn = dyn_cast<const llvm::Function>(D);
+  auto State = UpdateDigestAndGetGOInfo(GO, GOHash, GOIMap);
+  Changed = Changed || std::get<2>(State);
+
+  // Update the GO's hash using its contributions and dependencies.
+  auto &GOContributions = std::get<0>(State);
+  auto &GODependencies = std::get<1>(State);
+  for (const GlobalObject *const G : llvm::concat<const GlobalObject *const>(
+           make_range(GOContributions.begin(), GOContributions.end()),
+           make_range(GODependencies.begin(), GODependencies.end()))) {
+    const llvm::Function *const Fn = dyn_cast<const llvm::Function>(G);
     // if function will not be inlined, skip it
     if (Fn && Fn->hasFnAttribute(Attribute::NoInline))
       continue;
-    Changed = updateDigestUseCallDependencies(D, GOHash, Visited, GOIMap,
-                                              UpdateDigestAndGetDependencies) ||
+    Changed = updateDigestUseDependenciesAndContributions(
+                  G, GOHash, Visited, GOIMap, UpdateDigestAndGetGOInfo) ||
               Changed;
   }
+
   return Changed;
 }
 
+// Calculate the GOs' initial digest, dependencies, contributions and GONumber.
+std::pair<GOInfoMap, GONumber> caculateGONumAndGOIMap(Module &M) {
+  GOInfoMap GOIMap;
+  GONumber GONum;
+
+  for (auto &GV : M.globals()) {
+    if (GV.isDeclaration())
+      continue;
+    calculateInitialDigestAndDependenciesAndContributions(&GV, GOIMap);
+    ++GONum.VarNum;
+  }
+  for (auto &Fn : M.functions()) {
+    if (Fn.isDeclaration())
+      continue;
+    calculateInitialDigestAndDependenciesAndContributions(&Fn, GOIMap);
+    ++GONum.FuncNum;
+  }
+  return std::make_pair(std::move(GOIMap), std::move(GONum));
+}
+
 // Calculate the final GO hash by adding the initial hashes of its dependents
 // and the contributions and create the ticket metadata for GOs.
```

Comment:
Comment could explain what the tuple return value is about.

Reply:
I have deleted the comment since it has been given in the function declaration.